File: Python/LeetCodeTest/lc024_SwapNodesInPairs.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def swapPairs(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head
        count = 0
        output = None
        reverse_node = None
        connect_node = None
        while head is not None:
            if count % 2 == 1:
                if output is None:
                    output = head
                cache = head.next
                head.next = reverse_node
                reverse_node.next = cache
                if connect_node is not None:
                    connect_node.next = head
                head = reverse_node
                connect_node = reverse_node
            else:
                reverse_node = head
            count += 1
            head = head.next
        return output

if __name__ == "__main__":
    solution = Solution()
    node = ListNode(1)
    node.next = ListNode(2)
    node.next.next = ListNode(3)
    node.next.next.next = ListNode(4)
    result = solution.swapPairs(node)
    while result is not None:
        print(result.val)
        result = result.next

File: Java/src/com/liu/LeetCode/FifteenPatterns/LinkedListInPlaceReversal/LC024_SwapNodesInPairs.java
package com.liu.LeetCode.FifteenPatterns.LinkedListInPlaceReversal;

import com.liu.LeetCode.common.bean.ListNode;
import com.liu.LeetCode.common.factory.NodeCreator;
import com.liu.LeetCode.common.tools.Utils.NODE_TYPE;
import com.liu.LeetCode.common.tools.Utils;

/**
 * The In-place Reversal of a LinkedList pattern reverses parts of a linked list without using extra space.
 * Use this pattern when you need to reverse sections of a linked list.
 * Sample Problem:
 * Reverse a sublist of a linked list from position m to n.
 * Example:
 * Input: head = [1, 2, 3, 4, 5], m = 2, n = 4
 * Output: [1, 4, 3, 2, 5]
 * Explanation:
 * Identify the start and end of the sublist.
 * Reverse the nodes in place by adjusting the pointers.
 */

public class LC024_SwapNodesInPairs {

    private static ListNode swapPairsMy1(ListNode head) { // 两两交换，需要四个标记点
        if (head == null || head.next == null) {
            return head;
        }
        ListNode result = head.next;
        ListNode previous = null;
        ListNode current = head;
        ListNode next = head.next;
        while (current != null && next != null) {
            current.next = next.next;
            next.next = current;
            if (previous != null) {
                previous.next = next;
            }
            previous = current;
            current = current.next;
            if (current != null) {
                next = current.next;
            }
        }
        return result;
    }

    public static void main(String[] args) {
        int[][] valuesSet = new int[][]{
                {1, 2, 3, 4},
                {1, 2, 3},
                {1},
                {},
        };
        ListNode result;
        long startTime;
        long endTime;

        for (int i = 0; i < valuesSet.length; i++) {
            startTime = System.currentTimeMillis();
            ListNode node = (ListNode) NodeCreator.createNode(NODE_TYPE.NODE_TYPE_NORMAL_LIST, valuesSet[i]);
            result = swapPairsMy1(node);
            endTime = System.currentTimeMillis();
            System.out.println("swapPairs My1 result" + i + " " + Utils.getNodePrintableLog(NODE_TYPE.NODE_TYPE_NORMAL_LIST,
                    result) + " during time " + (endTime - startTime));
        }
    }
}
